agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/profiler/PerformanceMonitoringService.java

```diff
@@ -0,0 +1,164 @@
+    // Prevent double init
+    if (currentlyEnabled) {
+      return;
+    }
+    currentlyEnabled = true;
+    logger.warn("INITIALISING JFR PROFILING SUBSYSTEM THIS FEATURE IS IN BETA");
+
+    diagnosticEngine = null;
+    if (configuration.preview.profiler.enableDiagnostics) {
+      // Initialise diagnostic service
+      diagnosticEngine = startDiagnosticEngine();
+    }
+
+    alertServiceExecutorService =
+        Executors.newScheduledThreadPool(
+            2,
+            ThreadPoolUtils.createDaemonThreadFactory(
+                ProfilingInitializer.class, "ServiceProfilerAlertingService"));
+
+    profiler = new Profiler(configuration.preview.profiler, tempDir);
+
+    alerting =
+        AlertingSubsystemInit.create(
+            configuration,
+            TelemetryObservers.INSTANCE,
+            profiler,
+            telemetryClient,
+            diagnosticEngine,
+            alertServiceExecutorService);
+
+    uploadService =
+        new UploadService(
+            serviceProfilerClient,
+            builder -> {},
+            machineName,
+            processId,
+            telemetryClient::getAppId,
+            roleName);
+
+    try {
+      // Daemon remains alive permanently due to scheduling an update
+      profiler.initialize(uploadService, serviceProfilerExecutorService);
+
+    } catch (Throwable t) {
+      logger.error(
+          "Failed to initialise profiler service",
+          new RuntimeException(
+              "Unable to obtain JFR connection, this may indicate that your JVM does not"
+                  + " have JFR enabled. JFR profiling system will shutdown"));
+      alertServiceExecutorService.shutdown();
+    }
+  }
+
+  @Nullable
+  private DiagnosticEngine startDiagnosticEngine() {
+    try {
+      DiagnosticEngineFactory diagnosticEngineFactory = loadDiagnosticEngineFactory();
+      if (diagnosticEngineFactory != null) {
+        diagnosticEngineExecutorService =
+            Executors.newScheduledThreadPool(
+                1, ThreadPoolUtils.createNamedDaemonThreadFactory("DiagnosisThreadPool"));
+
+        DiagnosticEngine diagnosticEngine =
+            diagnosticEngineFactory.create(diagnosticEngineExecutorService);
+
+        if (diagnosticEngine != null) {
+          diagnosticEngine.init();
+        } else {
+          diagnosticEngineExecutorService.shutdown();
+        }
+        return diagnosticEngine;
+      } else {
+        logger.warn("No diagnostic engine implementation provided");
+      }
+    } catch (RuntimeException e) {
+      logger.error("Failed to load profiler factory", e);
+    }
+    return null;
+  }
+
+  private static DiagnosticEngineFactory loadDiagnosticEngineFactory() {
+    logger.info("loading DiagnosticEngineFactory");
+    return ServiceLoaderUtil.findServiceLoader(DiagnosticEngineFactory.class);
+  }
+
+  public void updateConfiguration(AlertingConfiguration alertingConfig) {
+    if (alerting != null) {
```

Comment:
does this method get called with null?

Reply:
should not no, just defensive coding

Reply:
oh sorry, just noticed, its actually the field that is null checked, not the argument, since alerting can be taken down when the profiler is disabled, it would be possible for this method to be called while alerting is null

Reply:
>  just noticed, its actually the field that is null checked, not the argument

oh yes I missed that too, thx